File: Example-ObjC/Simple/ViewController.m
//
//  ViewController.m
//  Simple
//
//  Created by Sedlak, Stefan on 7/21/16.
//  Copyright © 2016 Wirecard Technologies GmbH. All rights reserved.
//

#import "ViewController.h"

#import <PassKit/PassKit.h>
#import <PassKit/PKPaymentAuthorizationViewController.h>

#import <libextobjc/EXTScope.h>
#import <WDeCom/WDeCom.h>
#import <WDeComApplePay/WDeComApplePay.h>
#import <WDeComCard/WDeComCard.h>
#import <WDeComPayPal/WDeComPayPal.h>
#import <WDeComSEPA/WDeComSEPA.h>

#define AMOUNT [NSDecimalNumber decimalNumberWithMantissa:199 exponent:-2 isNegative:NO]

NSString *const PMTitleApplePay                 = @"ApplePay";
NSString *const PMTitleCard                     = @"Card";
NSString *const PMTitleCardManualBrandSelection = @"Card manual brand selection";
NSString *const PMTitlePayPal                   = @"PayPal";
NSString *const PMTitleSEPA                     = @"SEPA";

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UIButton *payBtn;

@end

@implementation ViewController

#pragma mark - Overriden

- (void)viewDidLoad {
    [super viewDidLoad];
    if (!self.client) {
        self.payBtn.enabled = NO;
    }
}

#pragma marks - Action

- (IBAction)onPay:(UIButton *)sender {
    UIAlertController *ac = [UIAlertController alertControllerWithTitle:@"Payment method"
                                                                message:nil
                                                         preferredStyle:UIAlertControllerStyleActionSheet];
    
    void (^handler)(UIAlertAction *action) = ^(UIAlertAction * _Nonnull action) {
        NSString *title = action.title;
        WDECPayment *payment = [self createPayment:title];
        if (!payment) {
            return;
        }
        sender.enabled = NO;
        @weakify(self, sender);
        [self.client makePayment:payment withCompletion:^(WDECPaymentResponse * _Nullable response, NSError * _Nullable error) {
            @strongify(self, sender);
            sender.enabled = YES;
            
            NSString *alertTitle = error ? @"Error" : @"Success";
            NSString *alertMessage = error ? error.localizedDescription : @"Item(s) has been purchased.";
            UIAlertController *ac = [UIAlertController alertControllerWithTitle:alertTitle
                                                                        message:alertMessage
                                                                 preferredStyle:UIAlertControllerStyleAlert];
            
            [ac addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
            [self presentViewController:ac animated:YES completion:nil];
        }];
    };
    
    BOOL canMakeApplePayPayment = [self canMakeApplePayPayment];
    if (canMakeApplePayPayment) {
        [ac addAction:[UIAlertAction actionWithTitle:PMTitleApplePay style:UIAlertActionStyleDefault handler:handler]];
    }
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleCard                        style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleCardManualBrandSelection    style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitlePayPal                      style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleSEPA                        style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:@"Cancel"                          style:UIAlertActionStyleCancel handler:handler]];
    
    [self presentViewController:ac animated:YES completion:nil];
}

#pragma mark - Private

- (BOOL)canMakeApplePayPayment
{
    NSArray<NSString *> *networks = @[PKPaymentNetworkMasterCard, PKPaymentNetworkVisa];
    BOOL result = [PKPaymentAuthorizationViewController canMakePaymentsUsingNetworks:networks];
    return result;
}

- (WDECPayment *)createPayment:(nonnull NSString *)title
{
    WDECPayment *result = nil;
    if ([title isEqualToString:PMTitleApplePay]) {
        result = [self createPaymentApplePay];
    } else if ([title isEqualToString:PMTitleCard]) {
        result = [self createPaymentCard];
        [[WDECCardLayout appearance] setManualCardBrandSelectionRequired:NO];
    } else if ([title isEqualToString:PMTitleCardManualBrandSelection]) {
        result = [self createPaymentCard];
        [[WDECCardLayout appearance] setManualCardBrandSelectionRequired:YES];
    } else if ([title isEqualToString:PMTitlePayPal]) {
        result = [self createPaymentPayPal];
    } else if ([title isEqualToString:PMTitleSEPA]) {
        result = [self createPaymentSEPA];
    }
    return result;
}

- (WDECPayment *)createPaymentApplePay
{
    static NSString *const APPLE_MERCHANT_ID = @"merchant.com.wirecard.paymentsdk.example.Simple";
    WDECApplePayManagedPayment *payment = [WDECApplePayManagedPayment new];
    payment.appleMerchantID = APPLE_MERCHANT_ID;
    payment.appleMerchantCountry = WDECCountryUS;
    payment.amount = AMOUNT;
    payment.currency = @"USD";
    payment.transactionType = WDECTransactionTypePurchase;
    
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    return payment;
}

- (WDECPayment *)createPaymentCard
{
    WDECCardPayment *payment = [WDECCardPayment new];
    payment.amount = AMOUNT;
    payment.currency = @"USD";
    payment.transactionType = WDECTransactionTypePurchase;
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"33f6d473-3036-4ca5-acb5-8c64dac862d1";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"9e0130f6-2e1e-4185-b0d5-dc69079c75cc";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    return payment;
}

- (WDECPayment *)createPaymentPayPal
{
    WDECOrderItem *orderItem = [WDECOrderItem new];
    orderItem.name = @"The Watch";
    orderItem.amount = AMOUNT;
    orderItem.amountCurrency = WDECCurrencyEUR;
    
    WDECOrder *order = [WDECOrder new];
    order.descriptor = @"DEMO DESCRIPTOR";
    order.number = [[NSUUID UUID] UUIDString];
    order.detail = @"DEMO ORDER DETAIL";
    order.items = @[orderItem];
    
    WDECPayPalPayment *payment = [WDECPayPalPayment new];
    payment.amount = AMOUNT;
    payment.currency = @"EUR";
    payment.transactionType = WDECTransactionTypeDebit;
    payment.order = order;
    
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"9abf05c1-c266-46ae-8eac-7f87ca97af28";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"5fca2a83-89ca-4f9e-8cf7-4ca74a02773f";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    payment.accountHolder = [self accountHolder];
    
    return payment;
}

- (WDECPayment *)createPaymentSEPA
{
    WDECSEPAPayment *payment = [WDECSEPAPayment new];
    payment.creditorID = @"DE98ZZZ09999999999";
    payment.mandateID = @"12345678";
    payment.amount = AMOUNT;
    payment.currency = @"EUR";
    payment.transactionType = WDECTransactionTypePendingDebit;
    
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"4c901196-eff7-411e-82a3-5ef6b6860d64";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"ecdf5990-0372-47cd-a55d-037dccfe9d25";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    return payment;
}

- (WDECCustomerData *)accountHolder {
    WDECCustomerData *accountHolder = nil;
    accountHolder = [WDECCustomerData new];
    accountHolder.lastName = @"Doe";
    
    return accountHolder;
}

@end

File: Example-Swift/Simple/ViewController.swift
//
//  ViewController.swift
//  Simple
//
//  Created by Vrana, Jozef on 2/21/17.
//  Copyright © 2017 Vrana, Jozef. All rights reserved.
//

import UIKit
import PassKit
import WDeComApplePay
import WDeCom
import WDeComCard
import WDeComPayPal
import WDeComSEPA

let AMOUNT = NSDecimalNumber.init(mantissa: 199, exponent: -2, isNegative: false)

let PMTitleApplePay = "ApplePay"
let PMTitleCard     = "Card"
let PMTitleCardManualBrandSelection = "Card manual brand selection"
let PMTitlePayPal   = "PayPal"
let PMTitleSEPA     = "SEPA"

class ViewController: PaymemtVC, UIActionSheetDelegate, UIPopoverPresentationControllerDelegate {
    
    @IBOutlet weak var payButton: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    @IBAction func onPay(_ sender: UIButton) {
    
        let actionSheetController = UIAlertController(title: "Payment", message: nil, preferredStyle: .actionSheet)
        
        func handler(actionTarget: UIAlertAction){
            
            
            if actionTarget.style == .cancel {
                self.dismiss(animated: true, completion: nil)
                return
            }
            
            let paymentTitle = actionTarget.title
            let payment = self.createPayment(title: paymentTitle!)
            
            if self.client == nil {
                return
            }
           
            self.client!.make(payment, withCompletion:{(response: WDECPaymentResponse?,error: Error?) in
                let alertTitle = error != nil ? "Error" : "Success"
                let alertMessage = error != nil ? error!.localizedDescription : "Item(s) has been purchased."
                let ac = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .actionSheet)
                ac.popoverPresentationController?.sourceView = self.view
                ac.popoverPresentationController?.sourceRect = CGRect(x: self.view.bounds.width / 2.0,y: self.view.bounds.height,width: 1.0,height: 1.0)

                ac.addAction(UIAlertAction(title:"OK", style:.default, handler:nil));
                self.present(ac, animated:true, completion:nil)
            })
        }
        let canMakeApplePayPayment = self.canMakeApplePayPayment()
        if canMakeApplePayPayment {
            actionSheetController.addAction(UIAlertAction(title: PMTitleApplePay, style: .default, handler: handler))
        }
        actionSheetController.addAction(UIAlertAction(title: PMTitleCard, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitleCardManualBrandSelection, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitlePayPal, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitleSEPA, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: handler))
        actionSheetController.popoverPresentationController?.sourceView = self.view
        actionSheetController.popoverPresentationController?.sourceRect = CGRect(x: self.view.bounds.width / 2.0,y: self.view.bounds.height,width: 1.0,height: 1.0)
        
        self.present(actionSheetController, animated: true, completion: nil)
    }
    
    func prepareForPopoverPresentation(_ popoverPresentationController: UIPopoverPresentationController) {
        
    }
    
    func canMakeApplePayPayment() -> Bool {
        let networks = [PKPaymentNetwork.masterCard, PKPaymentNetwork.visa]
        let result = PKPaymentAuthorizationViewController.canMakePayments(usingNetworks: networks)
        return result;
    }
    
    func createPayment(title:String) -> WDECPayment {
        var result : WDECPayment
        switch title {
        case PMTitleApplePay : result = self.createPaymentApplePay()
        case PMTitleCard : result = self.createPaymentCard()
        let cardLayout =  WDECCardLayout.appearance()
        cardLayout.manualCardBrandSelectionRequired = false
        case PMTitleCardManualBrandSelection :
            result = self.createPaymentCard()
            let cardLayout =  WDECCardLayout.appearance()
            cardLayout.manualCardBrandSelectionRequired = true
        case PMTitlePayPal : result = self.createPaymentPayPal()
        case PMTitleSEPA : result = self.createPaymentSEPA()
        default : result = WDECPayment()
        }
        return result
    }
    
    func createPaymentApplePay() -> WDECPayment {
        
        let APPLE_MERCHANT_ID = "merchant.com.wirecard.paymentsdk.example.Simple"
        let payment = WDECApplePayManagedPayment()
        payment.appleMerchantID = APPLE_MERCHANT_ID
        payment.appleMerchantCountry = .US
        payment.amount = AMOUNT
        payment.currency = WDECCurrencyGetISOCode(WDECCurrency.USD)
        payment.transactionType = .purchase
        
        let WD_MERCHANT_ACCOUNT_ID = "";
        let WD_MERCHANT_SECRET_KEY = "";
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentCard() -> WDECPayment {
        let payment = WDECCardPayment()
        payment.amount = AMOUNT
        payment.currency = "USD"
        payment.transactionType = .purchase
        let WD_MERCHANT_ACCOUNT_ID = "33f6d473-3036-4ca5-acb5-8c64dac862d1"
        let WD_MERCHANT_SECRET_KEY = "9e0130f6-2e1e-4185-b0d5-dc69079c75cc"
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentPayPal() -> WDECPayment {
        let orderItem = WDECOrderItem()
        orderItem.name = "The Watch"
        orderItem.amount = AMOUNT;
        orderItem.amountCurrency = .EUR;
        
        let order = WDECOrder()
        order.descriptor = "DEMO DESCRIPTOR"
        order.number = NSUUID.init().uuidString
        order.detail = "DEMO ORDER DETAIL"
        order.items = [orderItem]
        
        let payment = WDECPayPalPayment()
        payment.amount = AMOUNT
        payment.currency = "EUR"
        payment.transactionType = .debit
        payment.order = order;
        
        let WD_MERCHANT_ACCOUNT_ID = "9abf05c1-c266-46ae-8eac-7f87ca97af28"
        let WD_MERCHANT_SECRET_KEY = "5fca2a83-89ca-4f9e-8cf7-4ca74a02773f"
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentSEPA() -> WDECPayment {
        let payment = WDECSEPAPayment()
        payment.creditorID = "DE98ZZZ09999999999"
        payment.mandateID = "12345678"
        payment.amount = AMOUNT
        payment.currency = "EUR"
        payment.transactionType = .pendingDebit
                
        let WD_MERCHANT_ACCOUNT_ID = "4c901196-eff7-411e-82a3-5ef6b6860d64"
        let WD_MERCHANT_SECRET_KEY = "ecdf5990-0372-47cd-a55d-037dccfe9d25"
        
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
}


